controllers/datafoundationclient.go

```diff
@@ -0,0 +1,226 @@
+	if !found {
+		r.Log.Error(
+			fmt.Errorf("missing field: onboardingTicket"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+
+	providerEndpoint, found := offering.Spec.Config["dataFoundationProviderEndpoint"]
+	if !found {
+		r.Log.Error(
+			fmt.Errorf("missing field: dataFoundationProviderEndpoint"),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+	if !isValid {
+		r.Log.Info("parsing ManagedFusionOffering Data Foundation Client spec failed")
+		return fmt.Errorf("invalid ManagedFusionOffering Data Foundation Client spec")
+	}
+	r.Log.Info("parsing ManagedFusionOffering Data Foundation Client spec completed successfuly")
+
+	r.dataFoundationClientSpec = dataFoundationClientSpec{
+		onboardingTicket: onboardingTicket,
+		providerEndpoint: providerEndpoint,
+	}
+	return nil
+}
+
+func (r *dataFoundationClientReconciler) reconcilePhases() (ctrl.Result, error) {
+	if !r.offering.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	} else {
+		if err := r.reconcileStorageClient(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileCSV(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileDefaultBlockStorageClassClaim(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileDefaultFSStorageClassClaim(); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+func (r *dataFoundationClientReconciler) reconcileStorageClient() error {
+	r.Log.Info("Reconciling StorageClient")
+
+	if r.dataFoundationClientSpec.providerEndpoint == "" {
+		return fmt.Errorf("dataFoundationProviderEndpoint is empty")
+	}
+
+	dfStorageProviderEndpointAsURL, err := url.ParseRequestURI(r.dataFoundationClientSpec.providerEndpoint)
+	if err != nil {
+		return err
+	}
+	if dfStorageProviderEndpointAsURL.Host == "" {
+		return fmt.Errorf("cannot retrieve host from %s, might be a malformed or partial URL", r.dataFoundationClientSpec.providerEndpoint)
+	}
+
+	if r.dataFoundationClientSpec.onboardingTicket == "" {
+		return fmt.Errorf("onboardingTicket is empty")
+
+	}
+	_, err = r.CreateOrUpdate(&r.storageClient, func() error {
+		if err := r.own(&r.storageClient, true); err != nil {
+			return err
+		}
+		r.storageClient.Spec.OnboardingTicket = r.dataFoundationClientSpec.onboardingTicket
+		r.storageClient.Spec.StorageProviderEndpoint = r.dataFoundationClientSpec.providerEndpoint
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("unable to create/update StorageClient: %v", err)
+	}
+	return nil
+}
+
+func (r *dataFoundationClientReconciler) reconcileCSV() error {
+	r.Log.Info("Reconciling OCS Client Operator CSV")
+
+	csv, err := r.getCSVByPrefix(ocsClientOperatorName)
+	if err != nil {
+		return fmt.Errorf("unable to get OCS client CSV: %v", err)
+	}
+	if err := r.own(csv, false); err != nil {
+		return err
```

Comment:
These statements are no longer wrapped in an error `CreateOrUpdate` or `GetAndUpdate`. This means that the error itself will not be wrapped with another error the brings the context. Please add context
```suggestion
		return fmt.Errorf("unable to set owner reference on OCS client csv : %v", err)
```

Reply:
Added the error context